src/Umbraco.Infrastructure/Persistence/Repositories/Implement/EntityRepositoryBase.cs

```diff
@@ -122,66 +112,101 @@ protected virtual RepositoryCachePolicyOptions DefaultOptions
             }
         }
 
+        /// <summary>
+        /// Get the entity id for the <see cref="TEntity"/>
+        /// </summary>
+        protected virtual TId GetEntityId(TEntity entity)
```

Comment:
I'm intrigued here. IEntity as it stands has a single `Id` property of type `int`. This means you're boxing the id.

Reply:
I'm intrigued too! Took me a bit to find the culprit but 5 years ago with a change I made https://github.com/umbraco/Umbraco-CMS/commit/5d965346786e9055257c66cc1813b53f077dc8c0 and then started investigating why. It's because trying to cast int to `TId` doesn't work because TId can be anything. So this was a hack to avoid opening pandoras "box" ( 😂 ) since this is a broken abstraction and could be fixed with `IEntity<TId>`. There's many other issues here and many old TODOs lying around (now wishing I didn't look here!). I made a start on refactoring but it's quite a lot of changes and would like to resolve some of the old TODOs lying around. I'll keep this as a branch for now and see how it goes there. None of it's pretty. Need to move on to what I'm supposed to be working on but will come back to it.